python/lsst/pipe/tasks/dcrAssembleCoadd.py

```diff
@@ -785,19 +798,71 @@ def calculateGain(self, modelIter):
 
         Parameters
         ----------
-        modelIter : `int`
-            The current iteration of forward modeling.
+        convergenceList : `list` of `float`
+            The quality of fit metric from each previous iteration.
+        gainList : `list` of `float`
+            The gains used in each previous iteration: appended with new gain
+            value. Gains are numbers between ``self.config.baseGain`` and 1.
 
         Returns
         -------
         gain : `float`
             Relative weight to give the new solution when updating the model.
             A value of 1.0 gives equal weight to both solutions.
+
+        Raises
+        ------
+        ValueError
+            If ``len(convergenceList) != len(gainList)+1``.
         """
-        if self.config.useProgressiveGain:
-            iterGain = np.log(modelIter)*self.config.baseGain if modelIter > 0 else self.config.baseGain
-            return max(self.config.baseGain, iterGain)
-        return self.config.baseGain
+        nIter = len(convergenceList)
+        if nIter != len(gainList) + 1:
+            raise ValueError("convergenceList must be one element longer than gainList.")
+
+        if self.config.baseGain is None:
+            # If ``baseGain`` is not set, calculate it from the number of DCR subfilters
+            # The more subfilters being modeled, the lower the gain should be.
+            baseGain = 1./(self.config.dcrNumSubfilters - 1)
+        else:
+            baseGain = self.config.baseGain
+
+        if self.config.useProgressiveGain and nIter > 2:
+            # To calculate the best gain to use, compare the past gains that have been used
+            # with the resulting convergences to estimate the best gain to use.
+            # Algorithmically, this is a Kalman filter.
+            # If forward modeling proceeds perfectly, the convergence metric should
+            # asymptotically approach a final value.
+            # We can estimate that value from the measured changes in convergence
+            # weighted by the gains used in each previous iteration.
+            estFinalConv = [((1 + gainList[i])*convergenceList[i + 1] - convergenceList[i])/gainList[i]
+                            for i in range(nIter - 1)]
+            # WORDS about why negative is bad
+            estFinalConv = np.array(estFinalConv)
+            estFinalConv[estFinalConv < 0] = 0
+            # Because the estimate may slowly change over time, only use the most recent measurements.
+            estFinalConv = np.median(estFinalConv[max(nIter - 5, 0):])
+            lastGain = gainList[nIter - 2]
```

Comment:
Why not `gainList[-1]` to make it obvious it's the last value of the list?

Similarly for the other two `convergenceList[-2]` and `convergenceList[-1]`?